Support L2-less devices with fast redirect

jschwinger233 opened the pull request:
This PR adds L2 header for skb on L2-less devices before redirecting to internal veth, and enables bpf-based host routing work with wireguard as well as L2-less interfaces.

Fixes: #15075 

```release-note
Support L2-less devices with fast forward (bpf-based host routing) 
```

Signed-off-by: Zhichuan Liang <gray.liang@isovalent.com>

jschwinger233 committed:
bpf: add l2 header for skb on the l2-less dev

This commit adds l2 header before handle_ipv4 calls ipv4_local_delivery,
in order to enable fast-redirect from l2-less dev (e.g. cilium_wg0) to
internal veth.

Fixes: #15075

Signed-off-by: Zhichuan Liang <gray.liang@isovalent.com>

jschwinger233 committed:
daemon: enable host routing for wireguard and l2-less devices

Originally the host routing mode would be set lagacy if wireguard is
enabled or any l2-less device exists. This commit lifts the limitation.

Fixes: #15075

Signed-off-by: Zhichuan Liang <gray.liang@isovalent.com>

jschwinger233 commented on bpf/lib/l3.h:
```diff
@@ -14,6 +14,36 @@
+		/* If the dest dev is also an L3 dev, we skip the L2 header settings.
+		 */
+        return 0;
+    }
```
indent

jschwinger233 committed:
bpf: add unit test for l3 to l2 traffic

This commits add unittest case for L3 skb fast redirecting to L2 device.

Signed-off-by: Zhichuan Liang <gray.liang@isovalent.com>

jschwinger233 commented on test/bpf_tests/bpf_test.go:
```diff
@@ -85,6 +85,9 @@ func TestBPF(t *testing.T) {
 		if !strings.HasSuffix(entry.Name(), ".o") {
 			continue
 		}
+		if entry.Name() != "tc_nodeport_l3_dev.o" {
```
remember to delete

julianwiedmann commented on bpf/bpf_host.c:
```diff
@@ -441,6 +441,8 @@ handle_ipv4(struct __ctx_buff *ctx, __u32 secctx,
 	void *data, *data_end;
 	struct iphdr *ip4;
+	bool l2_hdr_required = true;
+	int l3_off = ETH_HLEN;
```
We can reduce the scope here, and just define these variables inside the relevant section:

```
ep = lookup_ip4_endpoint(ip4);
if (ep) {
	bool l2_hdr_required = true;
	int l3_off = ETH_HLEN;

	...
}
```

jschwinger233 replied:
Compiler will complain "error: mixing declarations and code is a C99 extension [-Werror,-Wdeclaration-after-statement]" - -

julianwiedmann commented on bpf/bpf_host.c:
```diff
@@ -534,6 +536,19 @@ handle_ipv4(struct __ctx_buff *ctx, __u32 secctx,
+						      false, l3_off))
+				return DROP_INVALID;
+		}
+
```
Should we only include this section when `#ifdef ENABLE_HOST_ROUTING`?

I believe we also need the same handling in the IPv6 path.

jschwinger233 replied:
I think currently ENABLE_HOST_ROUTING can't go along with ipv6, can it? 
host routing requires bpf masquerading: https://github.com/cilium/cilium/blob/master/daemon/cmd/kube_proxy_replacement.go#L461
bpf masquerading isn't supported for ipv6: https://github.com/cilium/cilium/blob/master/daemon/cmd/daemon_main.go#L1413

julianwiedmann replied:
Ah of course - thanks! (it's actually in the works at the moment: https://github.com/cilium/cilium/pull/23165)

So either we add the IPv6 datapath code already today (knowing that it's unused). Or we add a check/TODO at the right location in the agent code, to make sure we don't forget doing so when support for IPv6 BPF-Masq is added.

julianwiedmann reviewed (changes requested):
Thank you @jschwinger233 ! Some comments below.

dylandreimerink commented on bpf/tests/tc_nodeport_l3_dev.c:
```diff
@@ -0,0 +1,216 @@
+	/* We are building an L3 skb which doesn't have L2 header, so in theory
+	 * we need to set ctx->protocol = bpf_ntohs(ETH_P_IP), but bpf verifier
+	 * doesn't allow us to do so, and kernel also doesn't handle an L3 skb
+	 * properly (see https://elixir.bootlin.com/linux/latest/source/net/bpf/test_run.c#L1156).
```
Nit: Switch the link for a permalink version.
```suggestion
	 * properly (see https://elixir.bootlin.com/linux/v6.2.1/source/net/bpf/test_run.c#L1156).
```

dylandreimerink reviewed (approved):
Very nice!

nebril commented on daemon/cmd/kube_proxy_replacement.go:
```diff
@@ -462,15 +470,19 @@ func finishKubeProxyReplacementInit() error {
-			msg = fmt.Sprintf("BPF host routing is currently not compatible with Wireguard (--%s).", option.EnableWireguard)
+		//
+		case !mac.HaveMACAddrs(option.Config.GetDevices()) || option.Config.EnableWireguard:
+			if kernelVersion, err := version.GetKernelVersion(); err == nil {
```
We don't handle the `err` returned here if it's not nil. Is there any way to log it?

pchaigno commented on daemon/cmd/kube_proxy_replacement.go:
```diff
@@ -34,9 +34,17 @@ import (
+var (
+	// Host routing on the L3 devices counts on a patch for bpf_skb_change_head: https://lore.kernel.org/bpf/20210519154743.2554771-2-joamaki@gmail.com/
+	// The patch is available as of kernel 5.13-rc4, so we simply check kernel version to guarantee that.
+	hasSkbChangeHeadPatch = versioncheck.MustCompile(">=5.13.0")
```
We don't usually rely on specific kernel versions because that breaks too easily (e.g., RHEL kernels). Instead we probe the specific kernel feature we need if possible, or something that was introduced at the same time (same patchset ideally, same release otherwise).

In this case, it's not a feature but a bugfix. I'm expecting it will be backported to all relevant LTS kernels (?). Do we even need to enable this conditionally? What happens if the kernel doesn't support it? How obvious is the failure today?

jschwinger233 replied:
> I'm expecting it will be backported to all relevant LTS kernels

This is correct, I just checked the latest LTS kernels, 4.14.307, 4.19.275,  5.4.234, 5.10.172, and all of them have included the patch.

> What happens if the kernel doesn't support it?

According to the patch commit message, without it, redirecting a packet from a L3 device such as wireguard to the veth peer device will cause skb->data to point to the middle of the IP header on entry to tcp_v4_rcv().

pchaigno replied:
So Cilium would likely drop those packets because they are malformed. Maybe it's fine to just enable this by default without any probing? We can ask in the community meeting if unsure.

borkmann replied:
> So Cilium would likely drop those packets because they are malformed. Maybe it's fine to just enable this by default without any probing? We can ask in the community meeting if unsure.

I think if we document this with regards to stable kernel requirements, this would be fine too if pkts would be dropped for the case where they are not present. At least it would be a strong enough signal.

borkmann replied:
The other option is to probe for a helper which went into the same release as Jussi's fixes and only enable it then. This would ignore the stable releases for the time being, but at least it would ensure there are no gaps in between.

pchaigno replied:
I'm proposing dropping the probe mostly because it's fairly complicated to implement (for a probe) and this feels like a fairly corner case scenario already.

jschwinger233 replied:
Sure. To implement a probe looks fun, I could do it in future if really necessary.

jschwinger233 replied:
I'm going to resolve this conversation by now. The conclusion so far is to make no check and to presume costumers are using TLS kernels which have been backported with necessary patches. If it turns out to be needed in future, I can implement a feature probe then.

julianwiedmann replied:
Just for documentation purposes:

- `bpf: Set mac_len in bpf_skb_change_head` landed in kernel `5.13`
- bpf-host-routing is available since kernel 5.10
- between 5.10 and 5.13, the only kernel still maintained on kernel.org is `5.10`.
- `bpf: Set mac_len in bpf_skb_change_head` was backported to `5.10.42`

pchaigno commented on pkg/datapath/linux/config/config.go:
```diff
@@ -205,6 +205,7 @@ func (h *HeaderfileWriter) WriteNodeConfig(w io.Writer, cfg *datapath.LocalNodeC
 	cDefinesMap["LB4_BACKEND_MAP"] = "cilium_lb4_backends_v3"
 	cDefinesMap["LB4_REVERSE_NAT_SK_MAP"] = lbmap.SockRevNat4MapName
 	cDefinesMap["LB4_REVERSE_NAT_SK_MAP_SIZE"] = fmt.Sprintf("%d", lbmap.MaxSockRevNat4MapEntries)
+	cDefinesMap["IS_L3_DEV(ifindex)"] = "false"
```
Why is this needed? That macro is already defined lower in the same function.

jschwinger233 replied:
Removed, thanks for reminding.

pchaigno commented on bpf/node_config.h:
```diff
@@ -35,6 +35,10 @@ DEFINE_IPV6(HOST_IP, 0xbe, 0xef, 0x0, 0x0, 0x0, 0x0, 0x0, 0x1, 0x0, 0x0, 0xa, 0x
 
+#ifndef IS_L3_DEV
+#define IS_L3_DEV(ifindex) false
+#endif
```
This is already defined on line 213 of this same file.

jschwinger233 replied:
In the first place there wasn't such definition until I rebased the upstream. Let me delete the duplicate macro.

jschwinger233 replied:
The duplicated macro has been removed and I'll resolve this.

pchaigno reviewed (commented):
Nice work on your first datapath PR! :tada: 

I think the kernel version check needs to be changed to a kernel probe or even just removed. Besides that, just nit comments.

julianwiedmann commented on bpf/bpf_host.c:
```diff
@@ -256,9 +256,24 @@ handle_ipv6(struct __ctx_buff *ctx, __u32 secctx, const bool from_host,
 		/* Let through packets to the node-ip so they are
 		 * processed by the local ip stack.
 		 */
+		bool l2_hdr_required __maybe_unused = true;
```
Please move this declaration to the top of the block, so that the comment stays with the `ENDPOINT_F_HOST` handling.

julianwiedmann commented on bpf/bpf_host.c:
```diff
@@ -256,9 +256,24 @@ handle_ipv6(struct __ctx_buff *ctx, __u32 secctx, const bool from_host,
+			if (!____revalidate_data_pull(ctx, &data, &data_end,
+						      (void **)&ip6, sizeof(*ip6),
+						      false, l3_off))
+				return DROP_INVALID;
```
Don't think we need this revalidation for IPv6?

jschwinger233 replied:
Correct, since `ip6` is no more used afterwards.

borkmann replied:
And for the way out (L2 -> L3), this is being taken care of via maybe_add_l2_hdr in fib_redirect right?

jschwinger233 replied:
I think for the wireguard scenario where the typical return path is like lxc(L2) -> cilium_wg0(L3),  we don't have to take care of the extra L2 header in ctx, the helper `ctx_redirect(ctx, WG_IFINDEX, 0)` can redirect an L2 packet to L3 device, is that right?

julianwiedmann commented on bpf/bpf_host.c:
```diff
@@ -529,9 +544,27 @@ handle_ipv4(struct __ctx_buff *ctx, __u32 secctx,
 		/* Let through packets to the node-ip so they are processed by
 		 * the local ip stack.
 		 */
+		bool l2_hdr_required __maybe_unused = true;
+		int l3_off __maybe_unused = ETH_HLEN;
+
```
Same, please move in front of the comment.

julianwiedmann commented on bpf/lib/nodeport.h:
```diff
@@ -26,6 +26,7 @@
 #include "stubs.h"
 #include "proxy_hairpin.h"
 #include "fib.h"
+#include "neigh.h"
```
This is probably just leftover from pre-rebase, and can go away?

julianwiedmann reviewed (changes requested):
Just some smaller ~cosmetic things to pick up during the needed rebase, thanks!

borkmann commented on bpf/bpf_host.c:
```diff
@@ -538,12 +554,30 @@ handle_ipv4(struct __ctx_buff *ctx, __u32 secctx,
+						      false, l3_off))
+				return DROP_INVALID;
+		}
+#endif
```
Could this check be done unconditionally instead of via ifdefs?

jschwinger233 replied:
I think it could be done, since there's no side effect for normal L2-L2 redirect, or even host routing is off, this `maybe_add_l2_hdr` still does no harm. Julian, how do you like it? @julianwiedmann

julianwiedmann replied:
I don't see the advantage - if we only require it for host-routing, why burden the BPF program on eg. older kernels with the additional complexity?

jschwinger233 replied:
For the benefit of less code complexity, I decide to keep these `ifdef`. If any other thought coming into mind, please unresolve the conversation.

jschwinger233 committed:
dos: remove limitation part for wireguard documentation

The limitation for wireguard was originally recorded as not compatiable
with host-routing, and it's no more the case once pr #23935 is merged.

Signed-off-by: Zhichuan Liang <gray.liang@isovalent.com>

julianwiedmann commented on bpf/bpf_host.c:
```diff
@@ -538,12 +550,30 @@ handle_ipv4(struct __ctx_buff *ctx, __u32 secctx,
+		}
+#endif
+
 		return ipv4_local_delivery(ctx, ETH_HLEN, secctx, ip4, ep,
```
```suggestion
		return ipv4_local_delivery(ctx, l3_off, secctx, ip4, ep,
```

jschwinger233 replied:
Of course! Thanks!

julianwiedmann reviewed (changes requested):
Just one tiny fix-up needed. CI is all green already :heavy_check_mark:

julianwiedmann reviewed (approved):
Looks great, thank you @jschwinger233 ! :rocket:

qmonnet commented on bpf/tests/tc_nodeport_l3_dev.c:
```diff
@@ -0,0 +1,374 @@
+
+/* We wanted to tail call handle_policy from bpf_lxc, but at present it's
+ * impossible to #include both bpf_host.c and bpf_lxc.c at the same time.
+ * Therefore, we created a stud, mock_hanle_policy, to simply check if the
+ * our skb reaches there.
```
Not worth a respin, but if you need to update your PR for another reason:

```suggestion
 * Therefore, we created a stud, mock_hanle_policy, to simply check if
 * our skb reaches there.
```

There's another typo in the description of your first commit: `lagacy`.

qmonnet reviewed (approved):
Nice work!
